docs/compiling.rst

```diff
@@ -13,6 +15,129 @@ the [python_example]_ repository.
 
 .. [python_example] https://github.com/pybind/python_example
 
+A helper file is provided with pybind11 that can simplify usage with setuptools.
+
+To use pybind11 inside your ``setup.py``, you have to have some system to
+ensure that ``pybind11`` is installed when you build your package. There are
+four possible ways to do this, and pybind11 supports all four: You can ask all
+users to install pybind11 beforehand (bad), you can use
+:ref:`setup_helpers-pep518` (good, but very new and requires Pip 10),
+:ref:`setup_helpers-setup_requires` (discouraged by Python packagers now that
+PEP 518 is available, but it still works everywhere), or you can
+:ref:`setup_helpers-copy-manually` (always works but you have to manually sync
+your copy to get updates).
+
+An example of a ``setup.py`` using pybind11's helpers:
+
+.. code-block:: python
+
+    from setuptools import setup
+    from pybind11.setup_helpers import Pybind11Extension
+
+    ext_modules = [
+        Pybind11Extension(
+            "python_example",
+            ["src/main.cpp"],
+        ),
+    ]
+
+    setup(
+        ...,
+        ext_modules=ext_modules
+    )
+
+If you want to do an automatic search for the highest supported C++ standard,
+that is supported via a ``build_ext`` command override; it will only affect
```

Comment:
I vaguely remember asking this question before, why are we calling the the import `build_ext` rather than something more self-documenting? (``ChooseLatestCPPStandard`` -- ok, please don't call it that, but you get what I mean.)

Reply:
Because there is exactly one usage for it: `buildcmd={"build_ext": build_ext}` in `setup()`. We could call it anything we want, but this helps remind us that it is used for the `build_ext` command (it's triggered by any build process, but `python setup.py build_ext` does directly trigger this command). We could call it BuildExt if you'd like. I don't want to describe _what_ it does, because that could be augmented later; we could add support for non-default compilers (like mingw, for example), and things like that. In distutils, `build_ext` is the only command in which you can query the current compiler. You can't do it (directly) from extension. Because of that, many frameworks provide a `build_ext`.

Reply:
Pybind11BuildExt is another option. And `pybind11_build_ext`. You don't want someone to write `buildcmd={"build_ext": Pybind11BuildExt()} `, though, you pass the class not an instance. I like `build_ext` best, but am happy to use something else. Using the name `build_ext` is quite common, see NumPy, [for an example](https://github.com/numpy/numpy/blob/bd263912e11f7f79310c5516fab6bd0a7b7c8eb9/setup.py#L235).

Reply:
Pretty much everyone calls it `build_ext`, I think: https://github.com/mdavidsaver/setuptools_dso/blob/6b1b72f012bdba1c891ce7719dbd230345924209/src/setuptools_dso/dsocmd.py#L384